app.py

```diff
@@ -0,0 +1,224 @@
+from flask import Flask, render_template, redirect, request, session
+from hash import hash, check
+from psql import sql_select, sql_write,get_secret_key
+from service import get_card_list, get_questions, user_logged_in, fetch_data, reset_password, get_leaderboard, get_users, delete_user, log_out
+from psycopg2.extensions import AsIs 
+
+app = Flask(__name__)
+app.secret_key = get_secret_key().encode()
+
+@app.route('/')
+@app.route('/index')
+def index():
+    if user_logged_in():
+        return redirect('/home')
+    forgot = request.args.get('forgot', '')
+    stage = "1"
+    return render_template('index.html', forgot=forgot, stage=stage)
+
+@app.route('/index_action', methods=['POST'])
+def index_action():
+    forgot = request.form.get('forgot', '')
+    stage = request.form.get('stage', '')
+    email = request.form.get('email').lower()
+    answer = request.form.get('answer', '').lower()
+    query = sql_select('SELECT user_id, question, answer, username FROM users WHERE email=%s', email)
+    if query:
+        response = query[0]        
+        question = get_questions(response[1])        
+        msg = ''
+        if stage == "3":
+            if not check(answer, response[2]):                
+                stage = "2"
+                msg = 'Incorrect! Please try again...'
+        elif stage == "4":
+            user_id = query[0][0] 
+            new_password = hash(request.form.get('new_password'))
+            sql_write('UPDATE users SET password =%s WHERE user_id=%s', new_password, user_id)
+            msg = f"Password successfully updated for user '{response[3]}'"
+            return render_template('index.html', msg=msg)
+    else:
+        stage = "1"
+        question = ""
+        msg = 'Invalid email address'
+    return render_template('index.html', forgot=forgot, stage=stage, email=email, question=question, msg=msg)
+
+
+@app.route('/login_action', methods=['POST'])
+def login_action():
+    email = request.form.get('email', '').lower()
+    password = request.form.get('password', '')
+    query = sql_select('SELECT * FROM users WHERE email=%s', email)        
+    if query:
+        response = query[0]
+        if check(password, response[3]):
+            session['user_id'] = response[0]
+            session['username'] = response[1]
+            session['admin'] = response[6]
+            return redirect('/home')
+        else:
+            msg = "Incorrect username and/or password"
+        return render_template('index.html', msg=msg, query=query)
```

Comment:
query isn't clear what it is. You should try and use something more explicit, like user.